### ISSUE TITLE
Distillation loss

### ISSUE THREAD

**Issue by huu4ontocord** [lucidrains/marge-pytorch#7]
Could you please explain the distillation loss?

It looks like you are doing a comparison between cross attention and doc_similarity?

```
def` distill_attn_loss(evi_dots, doc_similarities, mask  = None, eps = 1e-5):
    evi_dots = rearrange(evi_dots, 'b l h i n j -> b (l h i) n j')

    if exists(mask):
        mask = rearrange(mask, 'b n j -> b () n j')
        evi_dots.masked_fill_(~mask, 0.)
        denom = mask.expand_as(evi_dots).sum(dim = (1, -1))
        evi_dots_mean = evi_dots.sum(dim = (1, -1)) / (denom + eps)
    else:
        evi_dots_mean = evi_dots.mean(dim = (1, -1))

    normed_evi_dots = evi_dots_mean.softmax(dim = -1)
    normed_evi_dots.detach_()

    doc_similarities = doc_similarities.softmax(dim = -1).log()
    loss = F.kl_div(doc_similarities, normed_evi_dots, reduction = 'batchmean')
    return loss
```

**Comment by lucidrains** [comment]
@ontocord this is actually an experimental feature from https://github.com/lucidrains/distilled-retriever-pytorch

is it working for you??

**Comment by huu4ontocord** [comment]
Thanks for the reference @lucidrains. I haven't tried yet but will let you know if I have results. Thanks!

**Comment by lucidrains** [comment]
@ontocord please do!
